pyfarm/agent/http/core/resource.py

```diff
@@ -129,90 +175,169 @@ def error(self, request, code, message):
             html_error = template.load("error.html")
-            deferred = html_error.render(
+            result = html_error.render(
                 code=code, code_msg=responses[code], message=message)
-            deferred.addCallback(request.write).addCallback(
-                lambda _: request.finish())
+            request.write(result)
 
-        elif "application/json" in content_types:
+        elif "application/json" in response_types:
             request.setResponseCode(code)
-            request.write({"error": message})
-            request.finish()
+            request.write(dumps({"error": message}))
 
         else:
             request.setResponseCode(UNSUPPORTED_MEDIA_TYPE)
-            request.write(
+            error = dumps(
                 {"error": "Can only handle text/html or application/json here"})
+            request.write(error)
+
+        request.finish()
+
+    def render_tuple(self, request, response):
+        """
+        Takes a response tuple of ``(body, code, headers)`` or
+        ``(body, code)`` and renders the resulting data onto
+        the request.
+        """
+        assert isinstance(response, (list, tuple)), type(response)
+        if len(response) == 3:
+            body, code, headers = response
+
+            if isinstance(headers, dict):
+                for header, value in headers.items():
+                    if isinstance(value, STRING_TYPES):
+                        value = [value]
+                    request.responseHeaders.setRawHeaders(header, value)
+
+            request.setResponseCode(code)
+            request.write(body)
             request.finish()
 
+        elif len(response) == 2:
+            body, code = response
+            request.setResponseCode(code)
+            request.write(body)
+            request.finish()
+
+        else:
+            self.error(
+                request, INTERNAL_SERVER_ERROR,
+                "Expected two or three length tuple for response"
+            )
+
+    @inlineCallbacks
+    def render_deferred(self, request, deferred):
+        """
+        An inline callback used to unpack a deferred
+        response object.
+        """
+        assert isinstance(deferred, Deferred)
+        response = yield deferred
+        self.render_tuple(request, response)
+
     def render(self, request):
-        # make sure that the requested content type is supported
-        content_type = self.content_types(request, default=["text/html",
-                                                            "application/json"])
-        if not self.CONTENT_TYPES & content_type:
+        assert self.SCHEMAS is not NotImplemented
+        assert self.ALLOWED_ACCEPT is not NotImplemented
+        assert self.ALLOWED_CONTENT_TYPE is not NotImplemented
+
+        # Ensure we can handle the content of the request
+        content_types = \
+            self.get_content_type(request) & self.ALLOWED_CONTENT_TYPE
+        if not content_types:
```

Comment:
This check must not be done if the incoming request does not have a body, which is the norm for GET, DELETE, and HEADER requests.